derp/derphttp/derphttp_test.go

```diff
@@ -206,3 +211,60 @@ func TestPing(t *testing.T) {
 		t.Fatalf("Ping: %v", err)
 	}
 }
+
+func TestHTTP2WebSocketFallback(t *testing.T) {
+	serverPrivateKey := key.NewNode()
+	s := derp.NewServer(serverPrivateKey, t.Logf)
+	defer s.Close()
+
+	httpsrv := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		up := r.Header.Get("Upgrade")
+		if up != "websocket" {
+			Handler(s).ServeHTTP(w, r)
+			return
+		}
+
+		c, err := websocket.Accept(w, r, &websocket.AcceptOptions{})
+		if err != nil {
+			t.Errorf("websocket.Accept: %v", err)
+			return
+		}
+		defer c.Close(websocket.StatusInternalError, "closing")
+		wc := wsconn.NetConn(r.Context(), c, websocket.MessageBinary)
```

Comment:
Nit: For clarity, `context.Background()` is preferable here because `r.Context()` is invalidated after `Accept`.

```suggestion
		wc := wsconn.NetConn(context.Background(), c, websocket.MessageBinary)
```

Reply:
Good point!